File: typescript/chapters/ch03/src/core/profiles.ts
// P03 在文件工具基础上增加 policy，作为外部副作用的执行前边界。
export type Capability = "loop" | "powershell" | "tool_registry" | "files" | "policy";

export interface ChapterProfile {
  // 固定章节号，组合根依此拒绝能力越级。
  readonly chapter: number;
  // 不可变能力白名单，决定本章可装配组件。
  readonly capabilities: ReadonlySet<Capability>;
}

class CapabilitySet implements ReadonlySet<Capability> {
  // 用不可导出底层 Set 的包装实现 ReadonlySet，防止 profile 能力被外部修改。
  readonly #values: Set<Capability>;

  constructor(values: readonly Capability[]) {
    this.#values = new Set(values);
  }

  // 返回能力数量，不暴露可变 Set。
  get size(): number {
    return this.#values.size;
  }

  // 判断能力是否属于当前章节白名单。
  has(value: Capability): boolean {
    return this.#values.has(value);
  }

  // 返回键值对迭代器以兼容 ReadonlySet。
  entries(): SetIterator<[Capability, Capability]> {
    return this.#values.entries();
  }

  // 返回能力名称迭代器。
  keys(): SetIterator<Capability> {
    return this.#values.keys();
  }

  // 返回能力值迭代器；Set 的键和值相同。
  values(): SetIterator<Capability> {
    return this.#values.values();
  }

  // 以原生 Set 回调签名遍历能力。
  forEach(
    callbackfn: (value: Capability, value2: Capability, set: ReadonlySet<Capability>) => void,
    thisArg?: unknown,
  ): void {
    this.#values.forEach((value) => {
      callbackfn.call(thisArg, value, value, this);
    });
  }

  // 支持 for...of 而不暴露写入入口。
  [Symbol.iterator](): SetIterator<Capability> {
    return this.values();
  }
}

export const P01: ChapterProfile = Object.freeze({
  chapter: 1,
  // P01 只有循环和 PowerShell，没有工具注册表，也没有文件能力。
  capabilities: new CapabilitySet(["loop", "powershell"]),
});

export const P02: ChapterProfile = Object.freeze({
  chapter: 2,
  // P02 在此增加工具注册表和文件能力，但还没有权限策略。
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files"]),
});

export const P03: ChapterProfile = Object.freeze({
  chapter: 3,
  // P03 增加 policy，bootstrap 检测到此能力后强制注入审批和审计。
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files", "policy"]),
});

// 返回模块内冻结单例，拒绝无效或尚未迁移的章节号。
export function profileForChapter(chapter: number): ChapterProfile {
  // 仅返回模块内冻结的单例，供组合根用引用相等性校验 profile 来源。
  if (chapter === 1) {
    return P01;
  }
  if (chapter === 2) {
    return P02;
  }
  if (chapter === 3) {
    return P03;
  }
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > 20) {
    throw new Error("chapter must be an integer from 1 to 20");
  }
  throw new Error(`Chapter ${chapter} has not been migrated to TypeScript yet`);
}

File: python/ch02_agent/agent_ch02/core/profiles.py
"""章节能力白名单。第 2 章只开放 loop 和 powershell。

这是什么：章节配置管理，定义每章可用的功能集合
Java 类比：类似配置常量类，定义 ChapterProfile DTO 和预定义常量
为什么需要：控制功能渐进开放，防止越章使用未讲解的特性

Java 对照：ChapterProfile 类似一个不可变配置 DTO，P01 类似 public static final 常量。
它防止第 2 章意外装配后续章节才应该出现的能力。
"""

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class ChapterProfile:
    """某个章节允许启用的能力集合。

    这是什么：章节配置的值对象，定义可用功能白名单
    Java 类比：类似 record ChapterProfile(int chapter, Set<String> capabilities)
    为什么需要：显式声明每章的功能边界，支持渐进式学习和测试隔离
    """

    chapter: int  # 章节编号，例如 1。
    capabilities: frozenset[str]  # 不可变集合，类似 Collections.unmodifiableSet。


P01 = ChapterProfile(1, frozenset({"loop", "powershell"}))  # 第一章固定能力配置。
P02 = ChapterProfile(2, frozenset({"loop", "powershell", "tool_registry", "files"}))  # 第二章增加工具注册和文件操作


def profile_for_chapter(chapter: int) -> ChapterProfile:
    """根据章节号取得配置；未迁移的章节明确报错。

    这是什么：章节配置查找函数
    Java 类比：类似 static ChapterProfile getProfile(int chapter) throws IllegalArgumentException
    为什么需要：根据章节号动态获取配置，对未实现章节快速失败
    """
    if chapter == 1:  # 第 1 章配置
        return P01
    if chapter == 2:  # 第 2 章配置
        return P02
    raise ValueError(f"第 {chapter} 章尚未迁移为 Python 版本")  # 未实现的章节明确报错
